Initial API spec for managed Cassandra project in CosmosDB

frrossms opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet. **March 1, 2021**
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet. **ASAP**
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [x] SDK of .NET (need service team to ensure code readiness)
      - [X] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [x] SDK of Go
      - [x] PowerShell
      - [X] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [X] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [X] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [X] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [X] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

frrossms commented:
Most of this was reviewed in the private repo (https://github.com/Azure/azure-rest-api-specs-pr/pull/2016/files) in the commit "Initial Swagger API for CosmosDB's Project Nova". The rest of the commit, all of which are small, were added since and need new review by ARM.

jianyexi commented:
@frrossms  For LintDiff,  it's because the below model's type can't be an array 
https://github.com/Azure/azure-rest-api-specs/blob/87bb2c506645ec393703daa831eab3aca97a56f3/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json#L919-L921

qbradley commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBManagedCassandraDataCenterGetInstanceView.json:
```diff
@@ -0,0 +1,17 @@
+    "202": {},
+    "200": {
+      "body": {
+        "isDataCenterReady": true
```
isDataCenterReady is an internal detail that shouldn't be included in the public api

frrossms replied:
Removed the data center instance view entirely.

varunkumta commented:
as per the breaking change policy - aka.ms/bcforapi, new apis require a new service/api version

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/providers/Microsoft.DocumentDB/cassandraClusters": {
```
>cassandraClusters [](start = 68, length = 17)

list calls at subscription level should use x-ms-pageable

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters": {
```
>cassandraClusters [](start = 103, length = 17)

as above, should be pageable

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/instanceView": {
+      "get": {
+        "operationId": "CassandraCluster_GetInstanceView",
+        "description": "Request that repair begin on this cluster as soon as possible.",
```
>Request that repair begin on this cluster as soon as possible [](start = 24, length = 61)

update

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/instanceView": {
```
>instanceView [](start = 135, length = 12)

this is the url pattern for a list call - if a cluster has only 1 instanceview you must model it as a /instanceViews/default api - and also support /instanceViews which will return the collection format with 1 resource

varunkumta replied:
if the response schema does not represent a resource (id, name, type, properties) - model it as a post instead with name fetchInstanceView

---
In reply to: [573443410](https://github.com/Azure/azure-rest-api-specs/pull/12861#discussion_r573443410) [](ancestors = 573443410)

frrossms replied:
Switch it to a POST of fetchInstanceView.

frrossms replied:
Actually, we're going with the GET pattern, so switched it to the new path.

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+          "items": {
+            "type": "object",
+            "properties": {
+              "id": {
```
>id [](start = 15, length = 2)

since this is returned for a list backups call - it must include name, id and type, timestamp must be in a custom property bag

frrossms replied:
Turned it into a proxy resource.

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+          }
+        ],
+        "responses": {
+          "202": {
```
get cannot be async

frrossms replied:
Made it just a 200.

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/backups": {
```
>backups [](start = 135, length = 7)

you must also support the named/single get i.e. /backups/{backupId}

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+          "properties": {
+            "dataCenterLocation": {
+              "type": "string",
+              "description": "(Can only be set at creation time.) The region this data center should be created in."
```
>Can only be set at creation time [](start = 31, length = 32)

use x-ms-mutability to specify this

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+        "properties": {
+          "description": "Properties of a managed Cassandra data center.",
+          "type": "object",
+          "properties": {
```
>properties [](start = 11, length = 10)

to support async creations with 201, you must also return a provisioningState property

frrossms replied:
Done.

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/dataCenters/{dataCenterName}/instanceView": {
```
>instanceVi [](start = 164, length = 10)

same comment as cluster instanceview

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+              "description": "(Read only.) The status of the Cosmos DB account at the time the operation was called. The status can be one of following. 'Creating' – the Cosmos DB account is being created. When an account is in Creating state, only properties that are specified as input for the Create Cosmos DB account operation are returned. 'Succeeded' – the Cosmos DB account is active for use. ...
+              "type": "string"
+            },
+            "restoreFromBackup": {
```
property name seems to imply a bool. since this is an id, consider restoreFromBackupId

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+              "type": "string",
+              "description": "Omit or set to null to create a new, empty cluster. If you are creating  new cluster by restoring a backup, then provide the unique resource identifier of the backup."
+            },
+            "delegatedManagementSubnetId": {
```
make sure to set x-ms-mutability on all properties that can only be set at creation time

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+              "enum": [
+                "None",
+                "Cassandra"
+              ]
```
also specify x-ms-enum

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+                "Cassandra"
+              ]
+            },
+            "initialCassandraAdminPassword": {
```
>initialCassandraAdminPassword [](start = 13, length = 29)

should be marked with x-ms-mutability and x-ms-secret [https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi013-secrets-must-be-annotated-with-x-ms-secret-and-x-ms-mutability](https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi013-secrets-must-be-annotated-with-x-ms-secret-and-x-ms-mutability)

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+              "type": "string",
+              "description": "Which version of Cassandra should this cluster converge to running (e.g., 3.11). When updated, the cluster may take some time to migrate to the new version."
+            },
+            "overrideClusterName": {
```
>overrideClusterName [](start = 13, length = 19)

again name seems to imply bool, consider clusterNameOverride instead

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+              "type": "boolean",
+              "description": "Should automatic repairs run on this cluster? If omitted, this is true, and should stay true unless you are running a hybrid cluster where you are already doing your own repairs."
+            },
+            "clientCertificates": {
```
consider grouping related properties into one object eg - the 3 cert related ones, 2 seedNode related ones, and 2 auth related ones

frrossms replied:
I had that, and the linters complained about it. Which do I believe?

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+                "enum": [
+                  "Up",
+                  "Down"
+                ]
```
x-ms-enum here and below

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBManagedCassandraClusterGet.json:
```diff
@@ -0,0 +1,73 @@
+          ],
+          "clientCertificates": [
+            {
+              "pem": "-----BEGIN CERTIFICATE-----\n...Base64 encoded certificate...\n-----END CERTIFICATE-----"
```
>"-----BEGIN CERTIFICATE [](start = 21, length = 23)

if these have private keys, then they should be annotated as secrets

frrossms replied:
They are public keys.

varunkumta commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBManagedCassandraDataCenterCreate.json:
```diff
@@ -0,0 +1,66 @@
+    "201": {
+      "body": {
+        "id": "/subscriptions/subid/resourceGroups/cassandra-prod-rg/providers/Microsoft.DocumentDB/cassandraClusters/cassandra-prod/dataCenters/dc1",
+        "name": "dc1",
```
type is also required

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+        },
+        "tags": {
+          "$ref": "#/definitions/Tags"
+        },
```
Let us define ClusterResource to only contain the actual Resource defined by us(no ARM properties). Let us define two new objects: ClusterResourceCreateUpdateProperties that will take in ClusterResource and let us define ClusterResourceGetResults object that will be returned when GET and LIST calls are made. 
Most of the APIs defined in CosmosDB follow this pattern. (You can find eamples for your reference)
These 5 ARM properties are defined in ARMResourceProperties, we can use that object here, instead of redefining them.

MehaKaushik replied:
Same comment for DataCenter resource

frrossms replied:
I was asked not to do this because want to be able to pipe from GET to PUT. I've switched to having them pull their properties from ARMResource and ARMProxyResource in types.json, though.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+            "provisioningState": {
+              "readOnly": true,
+              "description": "(Read only.) The status of the Cosmos DB account at the time the operation was called. The status can be one of following. 'Creating' – the Cosmos DB account is being created. When an account is in Creating state, only properties that are specified as input for the Create Cosmos DB account operation are returned. 'Succeeded' – the Cosmos DB account is active for use. ...
+              "type": "string"
```
>"type": "string" [](start = 14, length = 16)

Why is this not modeled as an Enum ?

frrossms replied:
Because we're not sure exactly what the values are aside from Succeeded and Failed.

frrossms replied:
I'm made it an enum for now.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/providers/Microsoft.DocumentDB/cassandraClusters": {
```
AFAIK, cassandraCluster is a tracked resource. Not sure if we support subscription move for it. If not we would need to. RPC here[[https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc003-tracked-resource-types-must-support-move](https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc003-tracked-resource-types-must-support-move)]

Another requirement for tracked resources is they must support Tags and those Tags must support Patch.

frrossms replied:
We don't yet, but I'll make it pretend to, and we'll just throw errors until we have it implemented.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+          "description": "PEM formatted public key.",
+          "type": "string"
+        }
+      }
```
I am not sure if we can include keys in a Resource body. Here is the ARM guideline [https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#impact-9](https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#impact-9)

frrossms replied:
These are public keys the customer needs to be able to get easily. They are not secrets.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+        "type": "string"
+      },
+      "description": "Tags are a list of key-value pairs that describe the resource. These tags can be used in viewing and grouping this resource (across resource groups). A maximum of 15 tags can be provided for a resource. Each tag must have a key no greater than 128 characters and value no greater than 256 characters. For example, the default experience for a template type is set with \"defaul ...
+    },
```
Is this the Tags defined by ARM or you created a new property called Tags for your scenario. If it is the former then we can resuse object definition from cosmosdb.json, if latter then maybe rename it to avoid confusion.

frrossms replied:
Tags defined by ARM. I've switch to using the ARMResource properties from cosmos-db.json.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+          "readOnly": true,
+          "type": "string",
+          "description": "The type of Azure resource."
+        },
```
ARMPRoyResource object defined in cosmosdb,json can be used. Similar comment like for ClusterResource here too.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/providers/Microsoft.DocumentDB/cassandraClusters": {
+      "get": {
```
>       [](start = 0, length = 6)

Tracked resources must also support SystemData. This is a new enforcement, our team is yet to make changes for globalDatabaseAccount for the same. You would also have to do this in very near future(hopefully in early Co2). [https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc027-systemdata-support](https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc027-systemdata-support)

frrossms replied:
Added it to clusters and data centers.

MehaKaushik replied:
I think only Clusters need this, Data center is not a tracked resource, afaik.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+            "schema": {
+              "$ref": "#/definitions/ClusterResource"
+            }
+          },
```
202 -> request is accepted and will complete async
200 -> OK, no-op (I think for our scenarios 200 and 202 can be used interchangeably) Although, I think 202 is more correct. 
201-> created 
Can you please make the descriptions a little more clear.

frrossms replied:
We can't return 200 when the request is completed, because a put is a request. We can use it for when the request is a no-op, and I've added that.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+            "description": "The repair request was successfully received.",
+            "schema": {
+              "$ref": "#/definitions/ClusterInstanceView"
+            }
```
200/202 on a GET call ? Looks like this GET is doing some action. Let us use POST instead ? 
Even the description is a little confusing.

MehaKaushik replied:
FYI: Post can be used for two designs: 
1. Performing an action on the resource
2.  Doing a list of keys

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+        ],
+        "responses": {
+          "202": {
+            "description": "Accepted. Update will proceed asynchronously."
```
Update on a "get" ?

MehaKaushik commented:
A general comment for overall PR is on the quality descriptions. These descriptions are the first thing that appears on a bing search.  Let us make them more formal and elaborate.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "#/definitions/DefaultErrorResponse"
+            }
```
Just FYI: In POST requests we can even return a response on completion.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+            "description": "The repair request was successfully received.",
+            "schema": {
+              "$ref": "#/definitions/ClusterInstanceView"
+            }
```
ClusterInstanceView is not modelled as a resource. In the URI instanceView should be plural and should be a list call. If is a point GET then it should have a param to provide the name of the instanceView. If InstanceView is always a single value then it uri should look like ....../instanceViews/default

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+                "format": "date-time",
+                "description": "The time this backup was taken, formatted like 2021-01-21T"
+              }
+            }
```
Same comment as for ClusterResource and DataCenter Resource.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/dataCenters": {
+      "get": {
+        "operationId": "CassandraDataCenter_ListDataCenters",
+        "description": "List all data centers in a particular managed Cassandra cluster.",
```
All list calls should have x-ms-pagable [https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi009-always-add-x-ms-pageable-to-list-calls](https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi009-always-add-x-ms-pageable-to-list-calls)

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+        ],
+        "responses": {
+          "202": {
+            "description": "Accepted. Deletion will run asynchronously."
```
>run [](start = 52, length = 3)

nit: Deletion will complete asynchronously

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+          "$ref": "cosmos-db.json#/definitions/ErrorResponse"
+        }
+      }
+    },
```
lets reuse this from cosmos-db.json

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1051 @@
+            "schema": {
+              "$ref": "#/definitions/ClusterResource"
+            }
+          },
```
I was thinking, can we make status codes for PUT and PATCH consistent. Both of them are individually correct, but why are they not same ? Any particular reason ?

frrossms replied:
Oversight. I've made them consistent.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/nova.json:
```diff
@@ -0,0 +1,1048 @@
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/fetchNodeStatus": {
+      "post": {
```
>   "post": { [](start = 3, length = 12)

Is this change made in our Service ? Ideally we should not merge swagger until the change at our Service side is deployed in Prod.

frrossms replied:
None of these exist in prod yet anyway. They're going out in hotfixes at the moment, and we're trying to coordinate all this stuff for our public preview date.

MehaKaushik commented:
> tag: package-2020-09

Please make the default apiVersion as 06-01-preview in all other readme's as well. With this change the SDK  PRs created would be generating SDK corresponding to your changes, you can review those sdks to see how your swagger changes are translating to code. 

---
Refers to: specification/cosmos-db/resource-manager/readme.md:30 in 97729a0. [](commit_id = 97729a0d1faf248e181b1edbeb3d0a72f00b9fce, deletion_comment = False)

piyengarr commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBManagedCassandraDataCenterCreate.json:
```diff
@@ -0,0 +1,55 @@
+    "clusterName": "cassandra-prod",
+    "dataCenterName": "dc1",
+    "body": {
+      "properties": {
```
We need to pass the name here. In the code we are throwing an exception, if the datacenter name in body does not match the URI

frrossms replied:
I think Quetzal was changing this?

piyengarr commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBManagedCassandraDataCenterGet.json:
```diff
@@ -0,0 +1,34 @@
+        "id": "/subscriptions/subid/resourceGroups/cassandra-prod-rg/providers/Microsoft.DocumentDB/cassandraClusters/cassandra-prod/dataCenters/dc1",
+        "name": "dc1",
+        "type": "Microsoft.DocumentDB/cassandraClusters/dataCenters",
+        "properties": {
```
The properties also contain "base64EncodedCassandraYamlFragment"

frrossms replied:
Good catch!

frrossms committed:
Move to a new API version.

frrossms committed:
Add nova.json

zhenglaizhang commented:
/azp run

MehaKaushik commented on specification/cosmos-db/resource-manager/readme.md:
````diff
@@ -30,6 +30,21 @@ openapi-type: arm
 tag: package-2021-01
 ```
 
+These settings apply only when `--tag=package-2021-03-preview` is specified on the command line.
````
Can you make this default tag everywhere. I suspect that would pass the CI.

frrossms replied:
Changed the `tag: package-2021-01` on line 30 to `tag: package-2021-03-preview`.

mentat9 commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1054 @@
+        "properties": {
+          "type": "object",
+          "properties": {
+            "timestamp": {
```
May want to consider adding systemData to this resource (already defined in common types). Details here: https://armwiki.azurewebsites.net/api_contracts/ResourceSystemData.html?q=SystemData

frrossms replied:
CosmosDB hasn't added it in general, so we're planning to add it at the same time that the rest of the resource provider gets it.

qbradley commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1058 @@
+        }
+      ],
+      "properties": {
+        "properties": {
```
ClusterResource is missing the prometheusEndpoint property

frrossms replied:
Added.

qbradley commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1058 @@
+        }
+      ],
+      "properties": {
+        "properties": {
```
DataCenterResource is missing the provisioningState property

frrossms replied:
Added.

qbradley commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1058 @@
+        "Deleting",
+        "Succeeded",
+        "Failed",
+        "Cancelled"
```
This should be Canceled with one el (see https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#provisioningstate-property)

qbradley commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/examples/CosmosDBManagedCassandraClusterListByResourceGroup.json:
```diff
@@ -0,0 +1,76 @@
+            "tags": {},
+            "properties": {
+              "provisioningState": "Succeeded",
+              "delegatedManagementSubnetId": "/subscriptions/536e130b-d7d6-4ac7-98a5-de20d69588d2/resourceGroups/customer-vnet-rg/providers/Microsoft.Network/virtualNetworks/customer-vnet/subnets/management",
```
the term "Management" means nothing to the user. As far as they are concerned, this is the subnet where the ip address of the prometheus endpoint appears.

I recommend calling this "delegatedPrometheusSubnetId"

frrossms replied:
After discussion within the group, we've decided not to change this, since prometheus will be optional but this is not, and we might end up adding another endpoint as well.

Commit:
added changes from stable apiVersion

ChenTanyi commented:
@frrossms Is your change of the API almost done? Since you had force-push, I don't know how many changes you make after ARM sign off. I would re-request ARM review after almost your change is done.
@PhoenixHe-msft @qiaozha Could you take a look at JS error? It seems a script bug.

colawwj commented:
@frrossms 
We found that the description in [restorable.json](https://github.com/frrossms/azure-rest-api-specs/blob/users/frrossms/nova/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/restorable.json#L267) exist a problem of "*/",It will be recognized as a comment symbol so that sdk package can not be run build,there are other five place has same problem in this restorable.json

frrossms commented:
@ChenTanyi I always force push, since I'm doing rebase -i to keep the commits clean. The changes are basically all getting the checks to pass.

@colawwj Thanks. The existing cosmos-db swagger seems pretty messed up. I'll go try to fix that.

frrossms commented:
@colawwj The */ is part of a specific string describing a permission users have to set for an API, so it can't be changed. This is a bug in the SDK generation script.

frrossms committed:
Add parameters to repair method on cassandraClusters.

frrossms committed:
Change fetchNodeStatus to be async POST

There appears to be no way to document the results of an async POST operation actually completing.

MehaKaushik commented:
> @frrossms
> We found that the description in [restorable.json](https://github.com/frrossms/azure-rest-api-specs/blob/users/frrossms/nova/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/restorable.json#L267) exist a problem of "*/",It will be recognized as a comment symbol so that JS sdk package can not be run build,there are other five place has same problem in this restorable.json

@kavskalyan can you please have a look.

vinhms commented:
> > @frrossms
> > We found that the description in [restorable.json](https://github.com/frrossms/azure-rest-api-specs/blob/users/frrossms/nova/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/restorable.json#L267) exist a problem of "*/",It will be recognized as a comment symbol so that JS sdk package can not be run build,there are other five place has same problem in this restorable.json
> 
> @kavskalyan can you please have a look.

if we can't fix the sdk generator, feel free to remove the "permission" string for now

frrossms commented:
I tried removing the */ sequences from restorable.json. It does not fix the JavaScript SDK generation. No motion on the filed incident: https://portal.microsofticm.com/imp/v3/incidents/details/229231273/home

How do I override this check?

frrossms committed:
Add 200 status code to fetchNodeStatus to see if it will make the Python SDK properly generate polling.

frrossms committed:
Fixed the fetchNodeStatus endpoint. It was actually missing x-ms-long-running-operation.

Removed the 200 status result I had put in before.

frrossms committed:
Add async response path for fetchNodeStatus.

frrossms committed:
Make linter happy.

soeom commented:
Can you please merge this change: https://github.com/Azure/azure-rest-api-specs/pull/13027

ChenTanyi commented:
> I tried removing the */ sequences from restorable.json. It does not fix the JavaScript SDK generation. No motion on the filed incident: https://portal.microsofticm.com/imp/v3/incidents/details/229231273/home
> 
> How do I override this check?

@colawwj @qiaozha @PhoenixHe-msft Could you take a look?

qbradley commented:
The last commit did not work for fetch node status.

Commit:
added changes from stable apiVersion

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1143 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "title": "Cosmos DB - Managed Cassandra",
```
Consider renaming the json file to something like ManagedCassandra,json

frrossms replied:
Done.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1143 @@
+    "/subscriptions/{subscriptionId}/providers/Microsoft.DocumentDB/cassandraClusters": {
+      "get": {
+        "operationId": "CassandraCluster_ListBySubscription",
+        "description": "List all managed Cassandra clusters in this subscription.",
```
nit: it should be CassandraClusters_* (plural) everywhere.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1143 @@
+          "200": {
+            "description": "Success. The repair operation will begin as soon as possible."
+          },
+          "409": {
```
>          "409":  [](start = 0, length = 17)

409 is conflict. 412 is precondition failed. This should be 412

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1143 @@
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/fetchNodeStatus/operationResults/{operationId}": {
+      "get": {
+        "operationId": "CassandraClusterFetchNodeStatusResults_Get",
+        "description": "Get the results of POSTing to the fetchNodeStatus method.",
```
>Get the results of POSTing to the fetchNodeStatus metho [](start = 24, length = 55)

nit: Not a user friendly description.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1143 @@
+          },
+          "204": {
+            "description": "No such cluster to delete."
+          },
```
nit: Cassandra cluster with the specified id does not exist.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1143 @@
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/repair": {
+      "post": {
+        "operationId": "CassandraCluster_RequestRepair",
+        "description": "Request that repair begin on this cluster as soon as possible.",
```
>as soon as possible. [](start = 66, length = 20)

Can this be quantified ?

frrossms replied:
Nope.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1143 @@
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/backups": {
+      "get": {
+        "operationId": "CassandraCluster_ListBackups",
+        "description": "List the backups of this cluster that are available to restore.",
```
>List the backups of this cluster that are available to restore. [](start = 24, length = 63)

Would be nice if you can describe what restore means here.

frrossms replied:
I'm pretty sure restoring a backup is a well understood term.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1143 @@
+            "description": "Accepted. Deletion will complete asynchronously."
+          },
+          "204": {
+            "description": "No such cluster to delete."
```
>"No such cluster to delete." [](start = 27, length = 28)

nit: Cassandra cluster with the specified id does not exist.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1143 @@
+            "schema": {
+              "$ref": "#/definitions/DataCenterResource"
+            }
+          },
```
Are you sure this is correct ?

frrossms replied:
Yes. In https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#put-resource it says:

"The response body should contain at least the original request that was PUT (and any other properties that would be returned in a GET, such as provisioningState, name, Id and type)."

MehaKaushik reviewed (approved):
Mostly looks good. Left some minor comments.

MehaKaushik commented:
> Can you please merge this change: #13027

@frrossms can you please add this change as well.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/nova.json:
```diff
@@ -0,0 +1,1143 @@
+          "200": {
+            "description": "Accepted. The data center will be updated asynchronously.",
+            "schema": {
+              "$ref": "#/definitions/DataCenterResource"
```
200 should not have a response with it. 

refer: [https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc015-put-apis-that-only-return-200](https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc015-put-apis-that-only-return-200)

frrossms replied:
It should have a response. Again, see https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#put-resource, which says that 200, 201, etc. should all return at least the properties that were posted. 200 just means no-op.

frrossms committed:
Rename nova.json to managedCassandra.json

frrossms committed:
Rename operationIds to be plural. Change status code for repair disabled to 412 (Precondition not met)

frrossms committed:
Fix javaScript SDK build (which is broken by have */ in descriptions)

frrossms committed:
Added lenient deduplication to make linter happy

frrossms commented:
@ChenTanyi Everything is passing, except it says the ApiDocPreview is still running here, but if you look at the actual pipeline (https://dev.azure.com/azure-sdk//internal/_build/results?buildId=753900&view=logs&j=f06b4322-f7bf-57ee-ecc5-6d0feac67556) it says that it succeeded. So I think this is ready to go.

Would you please merge?

frrossms committed:
Move fetchNodeStatus back since apparently the SDK generation handles this.

frrossms commented:
@ChenTanyi Okay, everything's passing and we're ready! (The ApiDocPreview pipeline is showing as ongoing here, but if you look at its actual details, it says it's done.) Can you please merge?

ChenTanyi commented:
@frrossms After the ARM re-signoff, it could be merged. Please remove the DoNotMerge Label.
@mentat9 Could you review again from ARM side? The PR has changed a lot after you sign off. Thanks!

frrossms commented:
To help with reviewing this, there are two layers of changes. One is merging the last preview and GA versions into the new preview version. That should be adequately checked by the breaking change linting, I believe. The others are actual changes to our new endpoints since @mentat9 signed off. I have extracted those changes as a patch with git diff:
[nova.patch.txt](https://github.com/Azure/azure-rest-api-specs/files/6047244/nova.patch.txt)

ChenTanyi commented:
@frrossms The diff looks good. But I saw you added another "nova.json". It may still need ARM to review.

frrossms commented:
If there's another nova.json, it's an accident and I need to delete it.

markcowl commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/cosmos-db.json:
```diff
@@ -0,0 +1,8956 @@
+      },
+      "required": [
+        "resource",
+        "options"
```
This is a breaking change from the previous stable api-version

frrossms replied:
Removed options from required list.

MehaKaushik replied:
Hmm, I had reverted this. I wonder if I missed this one line, but it appears in >10 places, where I fixed it. Thanks for catching this. 
Fred: can you please check if this is one place where I missed it or it happens throughout. If latter, then maybe my commits were over-written, if former, can you please remove options from required.

frrossms replied:
Checked. It was just the one place.

markcowl reviewed (changes requested):
This just markes breaking changes from the breaking change warnings in the PR

frrossms committed:
Fix inconsistencies between GA and preview previous API versions for cosmos-db.json.

frrossms commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/cosmos-db.json:
```diff
@@ -0,0 +1,8956 @@
+        "properties"
+      ]
+    },
+    
```
Fixed.

frrossms commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/cosmos-db.json:
```diff
@@ -0,0 +1,8956 @@
+      "required": [
+        "locations",
+        "databaseAccountOfferType",
+        "createMode"
```
It's a new field and is inherited from 2020-06-01-preview, so it's functionality that is not released yet and is being propagated forward into this preview version.

markcowl replied:
This is a breaking change from the stable rest api, and so needs a breaking change review

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/cosmos-db.json:
```diff
@@ -0,0 +1,8956 @@
+    "DatabaseAccountCreateUpdateProperties": {
+      "description": "Properties to create and update Azure Cosmos DB database accounts.",
+      "type": "object",
+      "discriminator": "createMode",
```
This is where our preview and stable apiVersions divert. This change currently exists only in preview.

markcowl replied:
This is a breaking change.  Breaking changes from stable require a breaking change review.

markcowl commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/managedCassandra.json:
```diff
@@ -0,0 +1,1102 @@
+          "200": {
+            "description": "Success. The repair operation will begin as soon as possible."
+          },
+          "412": {
```
So, we generally only want one response that is a failure, and that is the default response.  Any response code specifically modeled becomes a success in any generated clients.

markcowl replied:
If this is a new api, we should likely remove this.  If not, we want to avoid the breaking change.

frrossms replied:
Okay, going to remove it.

markcowl commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/managedCassandra.json:
```diff
@@ -0,0 +1,1102 @@
+          },
+          "200": {
+            "description": "Completed synchronously. This will only happen if the fields changed by patch require no changes to the actual Cassandra cluster, such as editing the tags."
+          },
```
No schema?  Generally this is the same schema as returned by Get and Put

frrossms replied:
Added.

markcowl commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/managedCassandra.json:
```diff
@@ -0,0 +1,1102 @@
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/dataCenters": {
+      "get": {
+        "operationId": "CassandraDataCenters_ListDataCenters",
```
Note: this can be CassandraDataCenters_List  this is not a required change, but a suggested change

markcowl replied:
Or alternately, CassandraClusters_ListDataCenters.  It depends if you want this grouped with data center operations ot cluster operations

frrossms replied:
Done

markcowl commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/managedCassandra.json:
```diff
@@ -0,0 +1,1102 @@
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/backups/{backupId}": {
+      "get": {
+        "operationId": "CassandraCluster_GetBackup",
```
```suggestion
        "operationId": "CassandraClusters_GetBackup",
```

frrossms replied:
D'oh. Fixed.

MehaKaushik commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/cosmos-db.json:
```diff
@@ -6825,8 +6825,7 @@
       "required": [
-        "resource",
-        "options"
+        "resource"
```
can you please check if this is consistent at other places in this file (for SQL, Cassandra, mongo, table, graph dbs and containers)

frrossms replied:
Just checked. It is consistent.

markcowl commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/managedCassandra.json:
```diff
@@ -0,0 +1,1102 @@
+          },
+          "200": {
+            "description": "Completed synchronously. This will only happen if the fields changed by patch require no changes to the actual Cassandra data center."
+          },
```
No schema?  Generally this should be the same as PUT

frrossms replied:
Added.

frrossms committed:
Add response body to PATCH for clusters and datacenters.

frrossms committed:
Add provisioningState, to dc examples, remove 412 code, fix a couple descriptions.

frrossms commented:
If I remove these fields, I'm breaking 2020-06-01-preview. If I add them, I'm breaking 2021-01-15. There is no way for me not to violate the breaking change policy at this point. Given that this is a new -preview release, keeping the fields is obviously the right thing to do.

There is no customer impact.

markcowl commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/cosmos-db.json:
```diff
@@ -0,0 +1,8959 @@
+            "schema": {
+              "$ref": "#/definitions/SqlStoredProcedureListResult"
+            }
+          }
```
no default response?

markcowl replied:
note: just commenting on this, this is not a required change. There are several instances throughout this file.  This is a correctness issue you guys should fix in a subsequent PR.  You may have S360 items for this

markcowl commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-03-01-preview/cosmos-db.json:
```diff
@@ -0,0 +1,8959 @@
+            "description": "The RUs per second of the MongoDB database update operation was completed successfully.",
+            "schema": {
+              "$ref": "#/definitions/ThroughputSettingsGetResults"
+            }
```
no default response?

markcowl replied:
note: just commenting on this, this is not a required change. There are several instances throughout this file. This is a correctness issue you guys should fix in a subsequent PR. You may have S360 items for this
